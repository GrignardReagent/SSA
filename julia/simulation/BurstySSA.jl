module BurstySSA

using Distributions: Geometric
using Catalyst
using JumpProcesses
using DifferentialEquations
using DataFrames
using Random
using Base.Threads

# symbolic registration must be evaluated at the top level 
@register_symbolic Geometric(b)

export simulate_bursty_model # this can be used directly when the module is imported

"""
    simulate_bursty_model(parameter_sets, time_points, ntrajectories; num_cores=nthreads())

Simulate mRNA counts from the 2-state bursty model via Gillespie SSA.

- `parameter_sets`: a Vector of Dicts with keys:
    "k", "b", "d", and optional "label".
- `time_points`: Vector of times at which to save M counts (e.g. `collect(0.0:1.0:100.0)`).
- `ntrajectories`: number of trajectories per parameter set.
- `num_cores`: number of threads to use (defaults to `Threads.nthreads()`).

Returns a `DataFrame` with columns: `:label, :time_t1, :time_t2, ...`.
"""
function simulate_bursty_model(parameter_sets, time_points, traj::Int; num_cores::Int=nthreads())
    
    @parameters b
    r = rand(Geometric(1/b)) + 1
    # the network
    bursty_model = @reaction_network bursty_model begin
        k, G --> G + $r*P
        d, P --> 0 
        end

    # initial molecule counts
    u0 = [:G => 1, :P => 0]
    tspan = (time_points[1], time_points[end])

    rows = Vector{Vector{Any}}()  # Any to allow String or Int labels

    @threads for ps in parameter_sets
        # parameters in the order rn.ps (Catalyst guarantees param order in rn.ps)
        params = Dict(
            :k => ps["k"],
            :b => ps["b"],
            :d => ps["d"],
        )

        label = Int32(get(ps, "label", 0))

        if traj == 1
            # single traj
            row = run_single(bursty_model, u0, tspan, params, time_points, label)
            push!(rows, row)
        else
            # multiple trajs, use ensemble
            ensemble_results = run_ensemble(bursty_model, u0, tspan, params, time_points, label, traj)
            append!(rows, ensemble_results)
        end
    end

    colsyms = [:label; Symbol.("time_" .* string.(time_points))]

    # Convert rows to a proper matrix format
    data_matrix = Matrix{Any}(undef, length(rows), length(colsyms))
    for (i, row) in enumerate(rows)
        data_matrix[i, :] = row
    end

    df = DataFrame(data_matrix, colsyms)
    return df
end

# One SSA trajectory, returns a row: [label, P(t1), P(t2), ...]
function run_single(bursty_model, u0, tspan, params, time_points, label)
    jinput = JumpInputs(bursty_model, u0, tspan, params)
    jprob  = JumpProblem(jinput; save_positions=(false, false)) # Disable saving at jump positions
    sol       = solve(jprob, SSAStepper(), saveat = time_points)
    p_counts  = Array(sol)[2, :]  # species order is [G, P], so we choose the 2nd row
    return [label; p_counts...]  # label + P counts
end

# run an ensemble of trajectories
function run_ensemble(bursty_model, u0, tspan, params, time_points, label, traj)
    jinput = JumpInputs(bursty_model, u0, tspan, params)
    jprob  = JumpProblem(jinput; save_positions=(false, false))
    eprob = EnsembleProblem(jprob)
    esol = solve(eprob, SSAStepper(); trajectories=traj, saveat=time_points)
    
    # Extract P counts for each trajectory
    ensemble_array = Array(esol)  # shape: (species, time_points, trajectories)
    p_counts_all = ensemble_array[2, :, :]  # P is the 2nd species
    
    # Create one row per trajectory
    rows = Vector{Vector{Any}}(undef, traj)
    for i in 1:traj
        p_counts = p_counts_all[:, i]  # P counts for trajectory i
        rows[i] = [label; p_counts...]
    end
    
    return rows
end

end # module