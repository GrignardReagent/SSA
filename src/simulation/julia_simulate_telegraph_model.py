# import the simulation functions
import multiprocessing
import pandas as pd 
import numpy as np
import os
os.environ.setdefault("PYTHON_JULIACALL_HANDLE_SIGNALS", "yes")

def simulate_telegraph_model(parameter_sets, time_points, size, num_cores=None):
    """
    Simulates the telegraph model using Julia implementation.
    Parameters:
        parameter_sets (list): List of parameter dictionaries for each system.
        time_points (numpy array): Array of time points for the simulation.
        size (int): Number of simulations per condition.
        num_cores (int, optional): Number of CPU cores to use. Defaults to all available cores.
    Returns:
        pd.DataFrame: DataFrame containing simulation results with columns for 'label' and time points.
    Usage:
        df_results = simulate_telegraph_model(parameter_sets, time_points, size, num_cores)
        # where parameter_sets is a list of dictionaries with keys like 'sigma_u', 'sigma_b', 'rho', 'd', and 'label'.
    """
    if num_cores is None:
        num_cores = multiprocessing.cpu_count()
    # print(f"Using {num_cores} cores for Julia simulation.")
    # threads need to be set using an environment variable before Julia starts
    os.environ["JULIA_NUM_THREADS"] = str(num_cores)
    
    from juliacall import Main as jl
    # Initialize Julia environment once (only if not already done)
    if not hasattr(simulate_telegraph_model, '_julia_initialized'):
        print("Initializing Julia environment...")
        jl.seval('using Pkg; Pkg.activate("/home/ianyang/stochastic_simulations/julia"); Pkg.instantiate()')
        jl.seval('using DataFrames, NPZ')
        jl.include("/home/ianyang/stochastic_simulations/julia/simulation/TelegraphSSA.jl")
        jl.seval('using .TelegraphSSA')
        simulate_telegraph_model._julia_initialized = True

    # Python → Julia conversion handled automatically
    jl.parameter_sets = parameter_sets
    jl.time_points = time_points
    
    # Run the simulation in Julia
    jl.seval(f'df_julia = simulate_telegraph_model(parameter_sets, time_points, {size})')
    
    # Convert Julia DataFrame directly to Python
    labels = np.array(jl.seval('Int64.(df_julia.label)'))
    counts_matrix = np.array(jl.seval('Int64.(Matrix(df_julia[:, Not(:label)]))'))
    
    # Create pandas DataFrame with same column format as Python version
    df_labels = pd.DataFrame(labels, columns=['label'])
    df_counts = pd.DataFrame(counts_matrix, columns=[f"time_{ti}" for ti in time_points])
    df = pd.concat([df_labels, df_counts], axis=1)

    # Check for NaN values before returning
    if df.isna().sum().sum() > 0:
        print("⚠️ Warning: NaN values detected in df!")
        print(df.isna().sum()[df.isna().sum() > 0])

    return df