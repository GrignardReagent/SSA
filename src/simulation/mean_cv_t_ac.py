#!/usr/bin/python

import numpy as np
from scipy.optimize import root_scalar
from utils.biological import check_biological_appropriateness
from stats.mean import calculate_mean_from_params
from stats.variance import calculate_variance_from_params
from stats.autocorrelation import calculate_ac_from_params
import warnings

# # Rescale parameters
# rho_tilda = rho / (sigma_b + sigma_u)
# d_tilda = d / (sigma_b + sigma_u)
# sigma_b_tilda = sigma_b / (sigma_b + sigma_u)
# sigma_u_tilda = sigma_u / (sigma_b + sigma_u)
# # t_ac_tilda is the only one that's multiplied by the scaling factor 
# t_ac_tilda = t_ac * (sigma_b + sigma_u)

# Compute rescaled parameters for a fixed ``sigma_sum``.
#
# This helper solves the mean, CV and autocorrelation time equations in the
# rescaled space assuming a known sum of switching rates ``sigma_sum`` (default = 1).
# It is intentionally kept private; the public :func:`find_tilda_parameters`
# routine wraps this function and automatically searches for an appropriate
# ``sigma_sum``.
def find_tilda_parameters(
    mu_target: float,
    t_ac_target: float,
    cv_target: float,
    sigma_sum: float = 1.0,
    ac_target: float = np.exp(-1),
    res_limit: float = 1e-3,
    lower: float = 1e-3,
    upper: float = 1e3,
    check_biological: bool = False,
    max_fano_factor: float = 20,
    min_fano_factor: float = 1.0,
    max_cv: float = 5,
):
    """Solve for ``rho``, ``d``, ``sigma_b`` and ``sigma_u`` given ``sigma_sum``.

    The equations are expressed in a rescaled space where all parameters are divided by ``sigma_sum`` to simplify the algebra.  

    Parameters
    ----------
    mu_target : float
        Target mean of the system.
    t_ac_target : float
        Target autocorrelation time.
    cv_target : float
        Desired coefficient of variation.
    sigma_sum : float, optional
        The sum ``sigma_b + sigma_u`` used to scale the system. Default to be 1.0. Changing this might lead to unexpected bahaviours!
    ac_target : float, optional
        Target autocorrelation value at ``t_ac_target``. Defaults to ``exp(-1)``.
    res_limit : float, optional
        Residual limit for root-finding. Defaults to ``1e-3``. This is used to check if the solution found is valid or not.
    lower : float, optional
        Lower bound for the root-finding search space for d_tilda. Defaults to ``1e-3``
    upper : float, optional
        Upper bound for the root-finding search space for d_tilda. Defaults to ``1e3``.
    check_biological : bool, optional
        If True, checks if the solution is biologically appropriate. Defaults to False.
    max_fano_factor : float, optional
        Maximum allowed Fano factor for the solution. Defaults to 20.
    min_fano_factor : float, optional
        Minimum allowed Fano factor for the solution. Defaults to 1.0.
    max_cv : float, optional
        Maximum allowed coefficient of variation for the solution. Defaults to 5.

    Returns
    -------
    tuple
        ``(rho, d, sigma_b, sigma_u)`` that reproduce the requested targets.

    Raises
    ------
    ValueError
        If no valid solution for ``d`` can be found.
    """
    
    # Basic feasibility checks
    if mu_target <= 0:
        raise ValueError("mu_target must be > 0.")
    if cv_target <= 0:
        raise ValueError("cv_target must be > 0.")
    if t_ac_target <= 0:
        raise ValueError("t_ac_target must be > 0 (AC(0)=1 carries no info).")
    if sigma_sum <= 0:
        raise ValueError("sigma_sum (Σ) must be > 0.")

    
    # using equation B, we find v via mu_target & cv_target^2
    v = (mu_target * (cv_target ** 2)) - 1
    
    # check if v > 0, if not, there's no solution
    if v <= 0:
        raise ValueError(f"Invalid parameters: v = (mu_target * cv_target ** 2) - 1 must be positive, got v = {v}. Reconsider mu_target and cv_target choices.")
    
    # t_ac_tilda = (sigma_b + sigma_u) * t; 
    t_ac_tilda = t_ac_target * sigma_sum
    
    # scaled lag: warn if far from 1 (conditioning)
    if t_ac_tilda < 1e-3 or t_ac_tilda > 1e3:
        warnings.warn(
            f"Scaled lag t̃ = Σ·t_ac = {t_ac_tilda:.3g} is ill-conditioned; "
            f"consider Σ≈1/t_ac_target so t̃≈1."
        )
    
    # find d_tilda from equation C, via t_ac_tilda, v
    def scaled_ac_equation(d_tilda):
        '''
        AC(t_ac_tilda), rescaled equation of AC(t) to find d_tilda:
        '''
        
        scaled_ACmRNA_eq = ((d_tilda * v * np.exp(- t_ac_tilda) + (d_tilda - v - 1) * np.exp(- d_tilda * t_ac_tilda)) / ((d_tilda -1) * (v + 1)))

        
        return float(scaled_ACmRNA_eq - ac_target)
    
    ################ Use root scalar to find d_tilda, AC(t_ac_tilda) - AC_target = 0  is a root-finding problem, so root_scalar is more appropriate ##################
    
    # the AC(t_ac_tilda) formula has the denominator (d_tilda -1)(v + 1), and d_tilda = 1 is undefined, so the search space needs to exclude 1.0; multiple decimal places to bracket away from 1.0
    candidates = [(lower, 0.999), (1.001, upper)] if (lower < 1.0 < upper) else [(lower, upper)]  # several decimal places are intentionally used here to avoid values close to 1.0
    
    d_tilda = None
    for a, b in candidates:
        try:
            result = root_scalar(scaled_ac_equation, bracket=[a, b],method='brentq', # Brent's method for root finding, generally considered the best for this. 
                                 xtol=1e-10, rtol=1e-10, maxiter=200)
            
            if result.converged:
                d_tilda_solution = result.root
                
                # error trap for no solution found for d_tilda_solution
                # 1. checks that absolute value of residue is less than res_limit
                # 2. checks that d_tilda_solution is not None
                # 3. checks that the value of d_tilda_solution is not close to the upper bounds plus/minus the absolute tolerance (atol)
                residue = scaled_ac_equation(d_tilda_solution)
                if abs(residue) > res_limit or d_tilda_solution is None or np.isclose(d_tilda_solution, b, atol=1e-4):
                    raise  ValueError(f"⚠️No valid solution found for parameter d_tilda: residual={residue:.2e}, d_tilda={d_tilda_solution:.4f}")
                
                d_tilda = d_tilda_solution
                
        except ValueError:
            pass
    if d_tilda is None: 
        raise ValueError(f'Could not find a valid solution for d_tilda using root_scalar method.')
    
    ################ Use root scalar to find d_tilda ##################
            
    # find rho_tilda from mu_target, d_tilda and v; by rearranging Equation A, definition of v in terms of tilda_params.
    #TODO: explanation of maths...
    rho_tilda = mu_target * d_tilda + (d_tilda + 1) * v
    
    # find sigma_b_tilda using solved parameters
    sigma_b_tilda = mu_target * d_tilda / (mu_target * d_tilda + (d_tilda + 1) * v)

    # find sigma_u_tilda using solved parameters
    sigma_u_tilda = (d_tilda + 1) * v / (mu_target * d_tilda + (d_tilda + 1) * v)
    
    # scaling back to original param space
    sigma_b = sigma_b_tilda * sigma_sum
    sigma_u = sigma_u_tilda * sigma_sum
    rho = rho_tilda * sigma_sum
    d = d_tilda * sigma_sum
    
    # optional biological check
    if check_biological:
        variance_target = (cv_target * mu_target) ** 2
        is_appropriate = check_biological_appropriateness(
            variance_target, mu_target, max_fano_factor, min_fano_factor, max_cv
        )
        if not is_appropriate:
            raise ValueError(
                "Solution is not biologically appropriate. "
                f"Consider adjusting target parameters: mu={mu_target}, cv={cv_target}, "
                f"which gives variance={variance_target:.2f}"
            )
        
    # back-substitution verification against original formulas
    mu_analytical = calculate_mean_from_params(rho, d, sigma_b, sigma_u)
    var_analytical = calculate_variance_from_params(rho, d, sigma_b, sigma_u)
    cv_analytical = np.sqrt(var_analytical) / mu_analytical
    ac_analytical = calculate_ac_from_params(rho, d, sigma_b, sigma_u, t_ac_target)
    
    # relative/absolute residuals
    rel = lambda x, y: abs(x - y) / max(1.0, abs(y)) # behaves like a relative error when abs(y) >= 1 and like an absolute error when abs(y) < 1. In contrastm, using just abs(y) in the denominator blows up (or becomes overly strict) when y is tiny or zero.
    mu_err = rel(mu_analytical, mu_target)
    cv_err = rel(cv_analytical, cv_target)
    ac_err = abs(ac_analytical - ac_target)
    if mu_err > res_limit or cv_err > res_limit or ac_err > res_limit:
        raise ValueError(
            "Back-substitution check failed: "
            f"μ_err={mu_err:.2e}, CV_err={cv_err:.2e}, AC_err={ac_err:.2e}. "
            "Try adjusting Σ so that t̃≈1, or relax targets."
        )
    return rho, d, sigma_b, sigma_u
