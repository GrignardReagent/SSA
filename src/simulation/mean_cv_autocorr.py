#!/usr/bin/python

import sympy as sp
from sympy import init_printing
import numpy as np
from scipy.optimize import fsolve, minimize_scalar
from simulation.mean_var_autocorr import equations

# Define symbols globally for reuse
rho, sigma_b, d, sigma_u, t, mu, sigma_sq, ac = sp.symbols('rho sigma_b d sigma_u t mu sigma_sq ac', real=True, positive=True)
init_printing(use_unicode=True)

################## EXPERIMENTAL, THIS IS NOT WORKING VERY WELL #####################
#  a quick version of find_parameters, to fix mean, autocorrelation and CV ONLY
def quick_find_parameters(sigma_b, mu_target=None, autocorr_target=None, cv_target=None):
    ''' Providing we are only fixing CV, mean and autocorrelation.
    Args:
        sigma_b (float): Value of sigma_b.
        mu_target (float, optional): Target mean.
        autocorr_target (float, optional): Target autocorrelation time.
        cv_target (float, optional): Target coefficient of variation.
    Returns:
        list: List of parameters (rho, sigma_u, d) that satisfy the equations.
    '''
    # Check if we have the required targets for this specialized solver
    if not (mu_target is not None and cv_target is not None and autocorr_target is not None):
        raise ValueError("quick_find_parameters requires mu_target, cv_target, and autocorr_target to be specified")
    
    # Step 1: Find d using numerical root finding with exponential transformation
    def d_equation(D):
        # Convert D to d using exponential transformation
        d_val = D
        
        # we use CV^2 to make calculations easier for the solvers
        cv_sq_target = cv_target ** 2
        # Use np.exp consistently for numerical calculations
        return -1/np.exp(1) + (-(d_val*mu_target*(1 + np.exp((autocorr_target*d_val*(d_val*mu_target - (-2 + mu_target + 2*cv_sq_target*mu_target)*sigma_b))/(d_val*mu_target + sigma_b - cv_sq_target*mu_target*sigma_b))*(-1 + cv_sq_target*mu_target))) + (-1 + cv_sq_target*mu_target**2 + cv_sq_target**2*mu_target**2 + np.exp((autocorr_target*d_val*(d_val*mu_target - (-2 + mu_target + 2*cv_sq_target*mu_target)*sigma_b))/(d_val*mu_target + sigma_b - cv_sq_target*mu_target*sigma_b))*(-1 + cv_sq_target*mu_target)**2)*sigma_b)/(cv_sq_target*np.exp(autocorr_target*d_val)*mu_target*(-(d_val*mu_target) + (-2 + mu_target + 2*cv_sq_target*mu_target)*sigma_b))
    
    def d_equation_sq_minus_sigma_u(D):
        '''This is f(d)^2 - sigma_u'''

        d_val = D
        
        # we use CV^2 to make calculations easier for the solvers
        cv_sq_target = cv_target ** 2
        
        sigma_u = -(((-1 + cv_sq_target*mu_target)*sigma_b*(d_val + sigma_b))/(-(d_val*mu_target) + (-1 + cv_sq_target*mu_target)*sigma_b))
        
        # Use np.exp consistently for numerical calculations
        return (-1/np.exp(1) + (-(d_val*mu_target*(1 + np.exp((autocorr_target*d_val*(d_val*mu_target - (-2 + mu_target + 2*cv_sq_target*mu_target)*sigma_b))/(d_val*mu_target + sigma_b - cv_sq_target*mu_target*sigma_b))*(-1 + cv_sq_target*mu_target))) + (-1 + cv_sq_target*mu_target**2 + cv_sq_target**2*mu_target**2 + np.exp((autocorr_target*d_val*(d_val*mu_target - (-2 + mu_target + 2*cv_sq_target*mu_target)*sigma_b))/(d_val*mu_target + sigma_b - cv_sq_target*mu_target*sigma_b))*(-1 + cv_sq_target*mu_target)**2)*sigma_b)/(cv_sq_target*np.exp(autocorr_target*d_val)*mu_target*(-(d_val*mu_target) + (-2 + mu_target + 2*cv_sq_target*mu_target)*sigma_b)))**2 - sigma_u
    
    # Square the equation for solutions where f(d) = 0
    d_equation_sq = lambda D: d_equation(D) ** 2
    
    # initialise system parameters
    D_guess = 1.0 / autocorr_target
    d_value = None
    rho = None
    sigma_u = None

    try:
        ################ Use minimization on d_equation_sq ##################
        result = minimize_scalar(d_equation_sq, method='bounded', bounds=(1e-3, 1e3))
        
        if result.success:
            D_solution = result.x
            d_value = D_solution
            # error trap for no solution found for d
            if result.fun > 1e-6 or d_value is None or d_value <= 0 or np.isclose(d_value, 1e3, atol=1e-4):
                # print(f'⚠️ WARNING: No solution found for d')
                raise ValueError("⚠️No valid solution found for parameter d")

            # Compute sigma_u safely
            cv_sq_target = cv_target ** 2
            sigma_u = -(((-1 + cv_sq_target*mu_target)*sigma_b*(d_value + sigma_b))/(-(d_value*mu_target) + (-1 + cv_sq_target*mu_target)*sigma_b))

            if sigma_u < 0:
                raise ValueError(f"Minimization led to negative sigma_u: {sigma_u:.4f}")

            # Compute rho from mean constraint
            rho = (d_value * mu_target * (sigma_b + sigma_u)) / sigma_b

            if rho <= 0:
                raise ValueError(f"Invalid rho computed: {rho:.4f}")
        ################ Use minimization on d_equation_sq ##################
    
            # Check if the solutions are valid (all parameters positive), if not, try the next guess
            if sigma_u <= 0 or rho <= 0:
                print(f"Warning: Invalid solution for D={D_guess}: rho={rho}, sigma_u={sigma_u}, d={d_value}")
                ########### MINIMIZATION ROUTINE ###########
                print("Trying bounded minimization...")
                result = minimize_scalar(d_equation_sq_minus_sigma_u, method='bounded', bounds=(1e-3, 1e3))

                if result.success:
                    D_solution = result.x
                    d_value = D_solution
                    
                    # error trap for no solution found for d
                    if result.fun > 1e-6 or d_value is None or d_value <= 0 or np.isclose(d_value, 1e3, atol=1e-4):
                        print(f'⚠️ WARNING: No solution found for d')
                        raise ValueError("No valid solution found for parameter d")

                    # Compute sigma_u safely
                    cv_sq_target = cv_target ** 2
                    sigma_u = -(((-1 + cv_sq_target*mu_target)*sigma_b*(d_value + sigma_b))/(-(d_value*mu_target) + (-1 + cv_sq_target*mu_target)*sigma_b))

                    if sigma_u < 0:
                        raise ValueError(f"Minimization led to negative sigma_u: {sigma_u:.4f}")

                    # Compute rho from mean constraint
                    rho = (d_value * mu_target * (sigma_b + sigma_u)) / sigma_b

                    if rho <= 0:
                        raise ValueError(f"Invalid rho computed: {rho:.4f}")
                else:
                    raise ValueError(f"Minimize_scalar routine failed: {result.message}")
                ########### MINIMIZATION ROUTINE ###########
            else:
                print(f"Found valid solution for D={D_guess}: rho={rho}, sigma_u={sigma_u}, d={d_value}")
    except Exception as e:
        print(f"Warning: Error during solving for D={D_guess}: {str(e)}")

    # DEBUG 
    # print(f"Final d_value: {d_value}, rho: {rho}, sigma_u: {sigma_u}")
    if d_value is None or rho is None or sigma_u is None:
        print(f"One of the following values is None: final d_value: {d_value}, rho: {rho}, sigma_u: {sigma_u}")
        try:
            ########### MINIMIZATION ROUTINE ###########
            print("Trying bounded minimization...")
            result = minimize_scalar(d_equation_sq_minus_sigma_u, method='bounded', bounds=(1e-3, 1e3))
            
            #DEBUG: Print the result of minimization
            print(f'res.fun for minimize_scalar: {result.fun}')

            if result.success:
                D_solution = result.x
                d_value = D_solution
                
                # error trap for no solution found for d
                if result.fun > 1e-6 or d_value is None or d_value <= 0 or np.isclose(d_value, 1e3, atol=1e-4):
                    print(f'⚠️ WARNING: No solution found for d')
                    raise ValueError("No valid solution found for parameter d")

                # Compute sigma_u safely
                cv_sq_target = cv_target ** 2
                sigma_u = -(((-1 + cv_sq_target*mu_target)*sigma_b*(d_value + sigma_b))/(-(d_value*mu_target) + (-1 + cv_sq_target*mu_target)*sigma_b))

                if sigma_u < 0:
                    raise ValueError(f"Minimization led to negative sigma_u: {sigma_u:.4f}")

                # Compute rho from mean constraint
                rho = (d_value * mu_target * (sigma_b + sigma_u)) / sigma_b

                if rho <= 0:
                    raise ValueError(f"Invalid rho computed: {rho:.4f}")
            else:
                raise ValueError(f"Minimize_scalar routine failed: {result.message}")
            ########### MINIMIZATION ROUTINE ###########
        except Exception as e:
            raise ValueError("Could not find a valid solution for parameter d")
        
    # Return the solution as a tuple
    solution = (rho, sigma_u, d_value)
    
    # Verify the solution by checking residuals
    residuals = equations(solution, sigma_b, mu_target, None, autocorr_target, cv_target, None)
    if all(abs(res) < 1e-4 for res in residuals):
        print(f"Found solution: rho={rho:.4f}, sigma_u={sigma_u:.4f}, d={d_value:.4f}")

        return solution
    else:
        print(f"Warning: Solution found but residuals are high: {residuals}")
        return solution
    
################## EXPERIMENTAL, THIS IS NOT WORKING VERY WELL #####################