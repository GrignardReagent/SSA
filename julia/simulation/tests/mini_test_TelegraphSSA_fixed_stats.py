#!/usr/bin/env python3

# import the simulation functions
import multiprocessing
import os
os.environ.setdefault("PYTHON_JULIACALL_HANDLE_SIGNALS", "yes")
num_cores = multiprocessing.cpu_count()
os.environ["JULIA_NUM_THREADS"] = str(num_cores)  # or set to desired number of threads
import numpy as np
import pandas as pd 
from simulation.mean_cv_t_ac import find_tilda_parameters
from simulation.julia_simulate_telegraph_model import simulate_telegraph_model
# plotting modules
import matplotlib.pyplot as plt
from visualisation.plots import plot_mRNA_dist

'''
Mini Test script (mini version of test_TelegraphSSA_fixed_stats.py) to replot distribution plots: testing whether the julia implementation of the telegraph SSA can produce trajectorie s matching a set of prescribed stats.
Python package modularised simulation function ``simulate_telegraph_model`` was used for simulation. 
'''

def run_simulation_test(mu_target, t_ac_target, cv_target, size=1000, output_dir = "mini_test_telegraph_fixed_stats_results"):
    """Run a single simulation test for given target statistics."""
    ###############################################################
    # Simulation
    ###############################################################
    print(f"\n{'='*60}")
    print(f"Testing: μ={mu_target}, t_ac={t_ac_target}, CV={cv_target}")
    print(f"{'='*60}")
    
    # Get the parameters using the scaled equations
    rho, d, sigma_b, sigma_u = find_tilda_parameters(mu_target, t_ac_target, cv_target)
    print(f"Parameters: ρ={rho:.6f}, d={d:.6f}, σ_b={sigma_b:.6f}, σ_u={sigma_u:.6f}")
    
    # Prepare parameter sets
    parameter_sets = [{
        'sigma_b': sigma_b,
        'sigma_u': sigma_u,
        'rho': rho,
        'd': d,
        'label': 0
    }]
    
    # Set time points - longer simulation for better autocorrelation estimation
    time_points = np.arange(0, 10_000, 1.0)
    
    # simulation
    df = simulate_telegraph_model(parameter_sets, time_points, size)
    
    ###############################################################
    # Plotting 
    ###############################################################
    # Create output directory if it doesn't exist
    os.makedirs(output_dir, exist_ok=True)
    
    # Plot data distribution
    fig, ax = plot_mRNA_dist(parameter_sets, df, kde=False)
    # Save figure with descriptive filename
    dist_filename = f"{output_dir}/telegraph_test_mu{mu_target}_tac{t_ac_target}_cv{cv_target}_dist.png"
    fig.savefig(dist_filename, dpi=300, bbox_inches='tight')
    plt.close(fig)  # Close to free memory
    print(f"Data distribution plot saved to: {dist_filename}")

    return {
        'mu_target': mu_target,
        't_ac_target': t_ac_target,
        'cv_target': cv_target,
        'parameters': {'rho': rho, 'd': d, 'sigma_b': sigma_b, 'sigma_u': sigma_u}
    }

###############################################################
# Main execution
###############################################################

# Define test targets
mu_targets = [10, 100]
t_ac_targets = [1, 5]
cv_targets = [0.5, 1.0]
size = 1000  # number of trajectories per simulation
output_dir = "mini_test_telegraph_fixed_stats_results"

# Run all combinations
results = []
total_tests = len(mu_targets) * len(t_ac_targets) * len(cv_targets)
current_test = 0

for mu_target in mu_targets:
    for t_ac_target in t_ac_targets:
        for cv_target in cv_targets:
            current_test += 1
            print(f"\nRunning test {current_test}/{total_tests}")
            
            try:
                result = run_simulation_test(mu_target, t_ac_target, cv_target, size=size, output_dir=output_dir)
                results.append(result)
            except Exception as e:
                print(f"Error in test μ={mu_target}, t_ac={t_ac_target}, CV={cv_target}: {e}")
                continue

# Create summary table
print(f"\n{'='*80}")
print("SUMMARY OF ALL TESTS")
print(f"{'='*80}")

summary_df = pd.DataFrame(results)
if not summary_df.empty:

    # Save summary to CSV
    summary_df.to_csv(f"{output_dir}/summary_results.csv", index=False)
    print(f"\nSummary saved to: {output_dir}/summary_results.csv")

else:
    print("No successful tests completed.")

print(f"\nAll tests completed. Results saved in '{output_dir}' directory.")
