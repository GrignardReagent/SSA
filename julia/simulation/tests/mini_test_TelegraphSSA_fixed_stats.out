/home/ianyang/micromamba/envs/stochastic_sim/lib/python3.11/site-packages/juliacall/__init__.py:61: UserWarning: torch was imported before juliacall. This may cause a segfault. To avoid this, import juliacall before importing torch. For updates, see https://github.com/pytorch/pytorch/issues/78829.
  warnings.warn(

Running test 1/8

============================================================
Testing: μ=10, t_ac=1, CV=0.5
============================================================
Parameters: ρ=30.250000, d=2.500000, σ_b=0.826446, σ_u=0.173554
Initializing Julia environment...
Data distribution plot saved to: mini_test_telegraph_fixed_stats_results/telegraph_test_mu10_tac1_cv0.5_dist.png

Running test 2/8

============================================================
Testing: μ=10, t_ac=1, CV=1.0
============================================================
Parameters: ρ=199.000000, d=10.000000, σ_b=0.502513, σ_u=0.497487
Data distribution plot saved to: mini_test_telegraph_fixed_stats_results/telegraph_test_mu10_tac1_cv1.0_dist.png

Running test 3/8

============================================================
Testing: μ=10, t_ac=5, CV=0.5
============================================================
Parameters: ρ=4.177016, d=0.232784, σ_b=0.557297, σ_u=0.442703
Data distribution plot saved to: mini_test_telegraph_fixed_stats_results/telegraph_test_mu10_tac5_cv0.5_dist.png

Running test 4/8

============================================================
Testing: μ=10, t_ac=5, CV=1.0
============================================================
Parameters: ρ=13.784452, d=0.251813, σ_b=0.182679, σ_u=0.817321
Data distribution plot saved to: mini_test_telegraph_fixed_stats_results/telegraph_test_mu10_tac5_cv1.0_dist.png

Running test 5/8

============================================================
Testing: μ=100, t_ac=1, CV=0.5
============================================================
Parameters: ρ=3124.000000, d=25.000000, σ_b=0.800256, σ_u=0.199744
Data distribution plot saved to: mini_test_telegraph_fixed_stats_results/telegraph_test_mu100_tac1_cv0.5_dist.png

Running test 6/8

============================================================
Testing: μ=100, t_ac=1, CV=1.0
============================================================
Parameters: ρ=19999.000000, d=100.000000, σ_b=0.500025, σ_u=0.499975
Data distribution plot saved to: mini_test_telegraph_fixed_stats_results/telegraph_test_mu100_tac1_cv1.0_dist.png

Running test 7/8

============================================================
Testing: μ=100, t_ac=5, CV=0.5
============================================================
Parameters: ρ=55.723698, d=0.255836, σ_b=0.459116, σ_u=0.540884
Data distribution plot saved to: mini_test_telegraph_fixed_stats_results/telegraph_test_mu100_tac5_cv0.5_dist.png

Running test 8/8

============================================================
Testing: μ=100, t_ac=5, CV=1.0
============================================================
Parameters: ρ=150.317101, d=0.257875, σ_b=0.171554, σ_u=0.828446
Data distribution plot saved to: mini_test_telegraph_fixed_stats_results/telegraph_test_mu100_tac5_cv1.0_dist.png

================================================================================
SUMMARY OF ALL TESTS
================================================================================

Summary saved to: mini_test_telegraph_fixed_stats_results/summary_results.csv

All tests completed. Results saved in 'mini_test_telegraph_fixed_stats_results' directory.
  Activating project at `~/stochastic_simulations/julia`
Using 12 threads for Julia simulation..
Using 12 threads for Julia simulation..
Using 12 threads for Julia simulation..
Using 12 threads for Julia simulation..
Using 12 threads for Julia simulation..
Using 12 threads for Julia simulation..
Using 12 threads for Julia simulation..
Using 12 threads for Julia simulation..
